### ISSUE TITLE
Connecting spaced out lines with dotted line

### ISSUE THREAD

**Issue by daredevil1302** [recharts/recharts#3394]
The issue which is not created via https://recharts.github.io/recharts-issue-helper will be closed immediately.

---

I have a yearly chart for all the months of the year. Now, for some months sometimes the data is missing (in this case the prices) and the price variable is `undefined`. This makes my chart look separated. For example there are lines for all months except february and may - > thats where the chart looks chopped up. The first picture represents what i want to achieve, and the second one what it looks like now. Does anyone have any advice with this? I tried adding new lines but doesn't seem to work

Here is my code: 

```
const calculateYearAndMonth = (data: LrpAnalysisPriceTrendsChartData[]) => {
  const yearCounter: Counter = {};
  const monthCounter: Counter = {};
  data.forEach((element) => {
    const year = element.name.slice(4, 8);
    const yearAndMonth = element.name;

    !yearCounter[year] ? (yearCounter[year] = 1) : (yearCounter[year] += 1);
    !monthCounter[yearAndMonth] ? (monthCounter[yearAndMonth] = 1) : (monthCounter[yearAndMonth] += 1);
  });
  return {
    yearCount: yearCounter,
    monthCount: monthCounter
  };
};

const tooltipRenderer =
  (selectedMeasure: string, unit: string, options: { id: string; name: string; color?: string }[]) =>
  ({ active, payload }: any) => {
    if (active && payload && payload[0]) {
      return (
        <TooltipContainer>
          <TooltipLabel>{payload[0].payload.endDate}</TooltipLabel>
          {payload.map(({ dataKey, color, value }: any) => {
            let yearInterval: string;
            dataKey === 'historicalPrice' ? (yearInterval = payload[0].payload.yearInterval) : (yearInterval = '');
            return (
              <TooltipRow key={dataKey} color={color}>
                {yearInterval} {selectedMeasure} :
                {` $${value.toFixed(2)}${unit === '$/cwt' ? '/cwt' : unit === '$/head' ? '/head' : ''}`}
              </TooltipRow>
            );
          })}
        </TooltipContainer>
      );
    }
  };

const renderYearTick = (data: LrpAnalysisPriceTrendsChartData[]) => (tickProps: any) => {
  const { x, y, payload } = tickProps;
  const counters: { [key: string]: { count: number; lastMonth: number } } = {};

  data.forEach((endorsementSummary) => {
    if (!counters[endorsementSummary['endMonth'].split('-')[1]]) {
      counters[endorsementSummary['endMonth'].split('-')[1]] = { count: 0, lastMonth: 0 };
      counters[endorsementSummary['endMonth'].split('-')[1]].count = 1;
      counters[endorsementSummary['endMonth'].split('-')[1]].lastMonth = Number(
        endorsementSummary['endMonth'].split('-')[0]
      );
    } else {
      counters[endorsementSummary['endMonth'].split('-')[1]].count += 1;
      counters[endorsementSummary['endMonth'].split('-')[1]].lastMonth = Number(
        endorsementSummary['endMonth'].split('-')[0]
      );
    }
  });

  const { value, offset } = payload;
  const currentMonth = Number(value.split('-')[0]);
  const centerCalculator = (t: number) => {
    switch (t) {
      case 1:
        return x + offset;
      case 2:
        return x;
      case 3:
        return x - offset;
      case 4:
        return x - offset * 2;
      case 5:
        return x - offset * 3;
      case 6:
        return x - offset * 4;
      case 7:
        return x - offset * 5;
      case 8:
        return x - offset * 6;
      case 9:
        return x - offset * 7;
      case 10:
        return x - offset * 8;
      case 11:
        return x - offset * 9;
      case 12:
        return x - offset * 10;
    }
  };

  if (currentMonth === counters[value.split('-')[1]].lastMonth) {
    return (
      <text x={centerCalculator(counters[value.split('-')[1]].count)} y={y} textAnchor="middle" fontSize={14}>{`${
        value.split('-')[1]
      }`}</text>
    );
  }
};

const renderLineTick = (data: LrpAnalysisPriceTrendsChartData[]) => (tickProps: any) => {
  const { x, y, payload, index } = tickProps;
  const counters: { [key: string]: number } = {};
  data.forEach((endorsementSummary) => {
    if (!counters[endorsementSummary['endMonth'].slice(2, 6)]) {
      counters[endorsementSummary['endMonth'].slice(2, 6)] = 1;
    } else {
      counters[endorsementSummary['endMonth'].slice(2, 6)] += 1;
    }
  });
  const { value, offset } = payload;
  const current_month = Number(value.split('-')[0]);
  const isLast = index === data.length - 1;

  if (index === 0 || isLast || current_month === 12) {
    const pathX = Math.floor(isLast || current_month === 12 ? x + offset * 2 : x);
    return <path d={`M${pathX},${y - 13}v${-35}`} stroke="#CCCCCC" />;
  }
  return null;
};

const renderMonthTick = (name: string) => name.slice(0, 3);
interface Props {
  data: LrpAnalysisPriceTrendsChartData[];
  width?: number;
  height?: number;
  gap: number;
  options: { id: string; name: string; color?: string }[];
  unit: number;
  selectedMeasure: { name: string; id: string };
  yearInterval: string;
}

const PriceTrendsChart: React.FC<Props> = ({
  width,
  height,
  data,
  options,
  gap,
  unit,
  selectedMeasure,
  yearInterval
}) => {
  const leftDomain = getDomain(data.map(({ currentPrice, historicalPrice }) => ({ currentPrice, historicalPrice })));
  const [range, setRange] = React.useState({ startIndex: 0, endIndex: data.length - 1 });

  React.useEffect(() => {
    setRange({ startIndex: 0, endIndex: data.length - 1 });
  }, [data.length]);

  const onBrushChange = (e: BrushRange) => {
    setRange(e);
  };
  console.log(data, 'podaci');
  const dataSliced = [...data.slice(range.startIndex, range.endIndex + 1)];
  const monthAndYearCount = calculateYearAndMonth(dataSliced);

  return (
    <>
      <ResponsiveContainer width="99%">
        <LineChart
          width={width}
          height={height}
          data={data}
          margin={{
            top: 5,
            right: 30,
            left: 30,
            bottom: 10
          }}
        >
          <YAxis
            label={{
              value: unit === Unit.CWT ? '$/cwt' : unit === Unit.HEAD ? '$/head' : '$',
              position: 'insideLeft',
              angle: -90,
              dy: 30,
              dx: -5,
              fontWeight: 200,
              fontSize: 15
            }}
            yAxisId="left"
            scale="linear"
            tickLine={false}
            tick={{ fontSize: 14 }}
            type="number"
            domain={leftDomain}
            tickFormatter={formatValue(2)}
            interval="preserveStart"
            tickCount={13}
            width={99}
          />
          <XAxis dataKey="name" height={40} tickFormatter={renderMonthTick} tick={{ fontSize: 12 }} />
          <XAxis
            dataKey="endMonth"
            axisLine={false}
            tickLine={false}
            interval={0}
            tick={renderLineTick(data)}
            height={30}
            scale="band"
            xAxisId="month"
          />
          {data[0].currentPrice !== 0 && data[0].endDate !== '' && (
            <XAxis
              dataKey="endMonth"
              axisLine={false}
              tickLine={false}
              interval={0}
              tick={renderYearTick(data)}
              height={35}
              scale="band"
              xAxisId="year"
            />
          )}
          <Tooltip
            content={tooltipRenderer(
              selectedMeasure.name,
              unit === Unit.CWT ? '$/cwt' : unit === Unit.HEAD ? '$/head' : '',
              options
            )}
          />
          <Legend margin={{ top: 20 }} />
          {console.log(data[2].currentPrice, 'trenutna cijena')}
          {data[0].currentPrice !== 0 && data[0].endDate !== '' ? (
             <Line
             key={'currentPrice'}
             name={selectedMeasure.name}
             type="linear"
             yAxisId={'left'}
             dataKey={'currentPrice'}
             stroke={'#0000FF'}
             strokeWidth={2}
             dot={false}
           />
          ) : (
            <Line
            key={'currentPrice'}
            name={selectedMeasure.name}
            type="linear"
            yAxisId={'right'}
            points={[{x: 12, y: 12, value: 240}]}
            strokeDasharray={"5 5"}
            dataKey={'historicalPrice'}
            stroke={'#0000FF'}
            strokeWidth={2}
            dot={false}
          />
          )}
          <Line
            key={'historicalPrice'}
            name={yearInterval + ' ' + selectedMeasure.name}
            type="linear"
            yAxisId={'left'}
            dataKey={'historicalPrice'}
            stroke={'#909090'}
            strokeWidth={2}
            dot={false}
          />
        </LineChart>
      </ResponsiveContainer>
    </>
  );
};
```

注意：不是用 https://recharts.github.io/recharts-issue-helper 创建的 issue 会被立即关闭。
![charTIdeal](https://user-images.githubusercontent.com/57237699/220639780-0ca3eab1-9610-4cd9-8e43-6b38e34193f9.png)
<img width="773" alt="chartPic" src="https://user-images.githubusercontent.com/57237699/220639787-8434a640-dfd3-45bc-b23e-7816c4b32e93.png">

**Comment by nikolasrieble** [comment]
The connectNulls property current is only allowing a boolean, see https://recharts.org/en-US/examples/LineChartConnectNulls

To enhance the API, we could extend the type such as 

```
connectNulls="4 1"
```
inspired by https://recharts.org/en-US/api/Line#strokeDasharray
Then, if the connectNulls is a string, the string will be parsed as a strokeDasharray pattern.

Does that make sense?

**Comment by daredevil1302** [comment]
> 

This makes sense. My client wants to see a dashed line where there is no data. connectNulls={true} allows me to connect but with a solid line. Is there any workaround that I could currently implement? @nikolasrieble

**Comment by nikolasrieble** [comment]
Sure, the workaround is to manipulate the data, and have two lines on top of each other. 
An underlying dashed line with connectNulls, and a second on top, without connectNulls. 

I did not validate that this works, but expect it to.

**Comment by daredevil1302** [comment]
This works!! Thank you @nikolasrieble 
<img width="845" alt="itworks" src="https://user-images.githubusercontent.com/57237699/220883632-a37ca54a-170c-4aae-8a25-9ec6c21dda79.png">

**Comment by daredevil1302** [comment]
@nikolasrieble  another issue occured now. When hovering with my mouse anywhere on the chart (provided my mouse always stays on the lines) my tooltip messages will keep duplicating and showing data that i dont want. It seems that when i hover with my mouse over the dotted line, it only registers the lower grayed out line, not the dotted one (which has a purpose to connect null/undefined values). If i remove the dotted line, the tooltip works perfectly fine.

TLDR: if i move my mouse on the lines non stop, tooltip message keeps multiplying.

I will attach a picture for clarity.
<img width="839" alt="problemNastao" src="https://user-images.githubusercontent.com/57237699/220964545-5aff9591-da69-4f65-a079-ba9bc290b986.png">

**Comment by ckifer** [comment]
Looks like you need a unique key on your tooltip items? `dataKey` isn't unique enough when you have two lines

**Comment by daredevil1302** [comment]
> Looks like you need a unique key on your tooltip items? `dataKey` isn't unique enough when you have two lines

Yes. That's because i want the dotted line to follow the same values as the solid line. I dont know what else to put in the `dataKey` prop of the dotted line

**Comment by ckifer** [comment]
Log the props and see if anything differs between the two. If you really have to check for a stroke-dasharray for the connectNulls variation and append some text to the key in that case

**Comment by daredevil1302** [comment]
Hmm, I don't think i understood what you meant by "append some text to the key in that case". If i change the `dataKey` of the stroke-dasharray line to something else, or append some text to it,  then it won't appear on the chart.

**Comment by ckifer** [comment]
Within your custom tooltip you call .map and render some elements. 

The duplication is because you are using dataKey as a React list key on the elements that you render there (this is outside of recharts, just a react thing). Whatever underlying element tooltip row is needs a unique key for that list. 

Take a look at this https://beta.reactjs.org/learn/rendering-lists
